datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -415,6 +441,74 @@ public void onPartitionMovement(Long notifyTimestamp) {
     _eventQueue.put(CoordinatorEvent.createPartitionMovementEvent(notifyTimestamp));
     _log.info("Coordinator::onPartitionMovement completed successfully");
   }
+
+  /**
+   * {@inheritDoc}
+   * Stop all the tasks and wait for new session to connect.
+   */
+  @Override
+  public void onSessionExpired() {
+    if (_shutdown) {
+      return;
+    }
+    stopEventThread();
```

Comment:
In the past we've seen some issues with handleLeaderDoAssignment() getting interrupted during cleanup and causing bad state such that we couldn't do future leader assignments anymore for BMM. This would require a full cluster restart to fix the state. We don't know the root cause yet, so we don't know how to  fix this scenario. Won't only calling stopEventThread() make such occurrences more frequent? If for some reason we cannot give more time to the eventThread to stop naturally, then how do we avoid that problem?

Have we added the relevant debug logs we needed for that issue (I recall it being an action item in one of the ZK session expiry documents that listed out ZK related issues)? If not, feel free to assign that item to me so that if it happens again I'll have something to debug with. Don't want to make that problem more frequent without appropriate logs in place. I've wasted 2-3 debugging sessions on it in the past to find out that it was related to early killing of handleLeaderDoAssignment, and after the timeout was increased, we stopped seeing it (so far). Which is why I haven't had a chance to root cause it since. Unit test don't help either.

Reply:
done.